PackedByteArray, Array slice end exclusive, rename PackedByteArray subarray to slice

nathanfranke opened the pull request:
- `slice` end index is now exclusive 
- `Packed*Array::subarray` is now `Packed*Array::slice`

```swift
func _ready() -> void:
	var array := PackedByteArray([0, 1, 2, 3, 4, 5, 6, 7])
	
	aprint(array.slice(0, 4))             ### 0, 1, 2, 3
	# Lines below would have crashed
	aprint(array.slice(4, 8))             ### 4, 5, 6, 7
	aprint(array.slice(0, array.size()))   ### 0, 1, 2, 3, 4, 5, 6, 7

func aprint(array: PackedByteArray) -> void:
	var s := String()
	for i in array:
		s += ", " + str(i)
	print(s.substr(2))
```

aaronfranke commented:
I'm curious why it was inclusive in the first place. This was added in #31172, pinging @creikey.

nathanwfranke commented:
Updated to new master branch. Should work completely fine but additional testing may be needed

nathanwfranke commented:
I am assuming I have to rebase this since PoolByteArray changed to PackedByteArray. I should be able to do it tonight

nathanfranke commented:
I updated this to master. I need to understand exactly what the expected behaviour should be, especially with negative steps.

Examples of current behaviour:
```py
var a = [0, 1, 2, 3, 4, 5, 6, 7]

print(a.slice(0, 4)) # [0, 1, 2, 3]
print(a.slice(4, 8, -1)) # []
print(a.slice(8, 0, -1)) # [7, 6, 5, 4, 3, 2, 1, 0]
print(a.slice(-1, 0, -1)) # [6, 5, 4, 3, 2, 1, 0]
```

nathanfranke commented:
Rebased again. Added documentation for `PackedColorArray`, `PackedFloat32Array`, `PackedFloat64Array`, `PackedInt32Array`, `PackedInt64Array`, `PackedStringArray`, `PackedVector2Array`, `PackedVector3Array`

Commit:
PackedByteArray, Array slice end exclusive, rename subarray to slice

akien-mga reviewed (approved):
Makes sense to me :+1:

mhilbrunner reviewed (approved):
PR meeting: makes sense, approved :)

akien-mga commented:
Thanks!

akien-mga commented:
See #56661 for a follow-up discussion on the behavior for negative `end`.
